Source/SpatialGDK/Private/Utils/ComponentFactory.cpp

```diff
@@ -0,0 +1,288 @@
+{
+	bool bWroteSomething = false;
+
+	// Populate the replicated data component updates from the replicated property changelist.
+	if (Changes.RepChanged.Num() > 0)
+	{
+		FChangelistIterator ChangelistIterator(Changes.RepChanged, 0);
+		FRepHandleIterator HandleIterator(ChangelistIterator, Changes.RepCmds, Changes.RepBaseHandleToCmdIndex, 0, 1, 0, Changes.RepCmds.Num() - 1);
+		while (HandleIterator.NextHandle())
+		{
+			const FRepLayoutCmd& Cmd = Changes.RepCmds[HandleIterator.CmdIndex];
+			const FRepParentCmd& Parent = Changes.Parents[Cmd.ParentIndex];
+
+			if (GetGroupFromCondition(Parent.Condition) == PropertyGroup)
+			{
+				const uint8* Data = Changes.SourceData + HandleIterator.ArrayOffset + Cmd.Offset;
+				TSet<const UObject*> UnresolvedObjects;
+
+				AddProperty(ComponentObject, HandleIterator.Handle, Cmd.Property, Data, UnresolvedObjects, ClearedIds);
+
+				if (UnresolvedObjects.Num() == 0)
+				{
+					bWroteSomething = true;
+				}
+				else
+				{
+					if (!bIsInitialData)
+					{
+						// We have to write something if it's initial data. ...I think. ?????????????????????
+						Schema_ClearField(ComponentObject, HandleIterator.Handle);
+					}
+
+					PendingUnresolvedObjectsMap.Add(HandleIterator.Handle, UnresolvedObjects);
+				}
+			}
+
+			if (Cmd.Type == REPCMD_DynamicArray)
+			{
+				if (!HandleIterator.JumpOverArray())
+				{
+					break;
+				}
+			}
+		}
+	}
+
+	// TODO: Handover
+	// Populate the handover data component update from the handover property changelist.
+	//for (uint16 ChangedHandle : Changes.HandoverChanged)
+	//{
+	//}
+
+	return bWroteSomething;
+}
+
+void ComponentFactory::AddProperty(Schema_Object* Object, Schema_FieldId Id, UProperty* Property, const uint8* Data, TSet<const UObject*>& UnresolvedObjects, TArray<Schema_FieldId>* ClearedIds)
+{
+	if (UStructProperty* StructProperty = Cast<UStructProperty>(Property))
+	{
+		UScriptStruct* Struct = StructProperty->Struct;
+		FSpatialNetBitWriter ValueDataWriter(PackageMap, UnresolvedObjects);
+		bool bHasUnmapped = false;
+
+		if (Struct->StructFlags & STRUCT_NetSerializeNative)
+		{
+			UScriptStruct::ICppStructOps* CppStructOps = Struct->GetCppStructOps();
+			check(CppStructOps); // else should not have STRUCT_NetSerializeNative
+			bool bSuccess = true;
+			if (!CppStructOps->NetSerialize(ValueDataWriter, PackageMap, bSuccess, const_cast<uint8*>(Data)))
+			{
+				bHasUnmapped = true;
+			}
+			checkf(bSuccess, TEXT("NetSerialize on %s failed."), *Struct->GetStructCPPName());
+		}
+		else
+		{
+			TSharedPtr<FRepLayout> RepLayout = NetDriver->GetStructRepLayout(Struct);
+
+			RepLayout_SerializePropertiesForStruct(*RepLayout, ValueDataWriter, PackageMap, const_cast<uint8*>(Data), bHasUnmapped);
+		}
+
+		Schema_AddPayload(Object, Id, ValueDataWriter);
```

Comment:
What does this do? Is this a `Schema_AddBytes` wrapper for `FBitWriter`?

Reply:
Correct.